drivers/serial/uart_bflb.c

```diff
@@ -0,0 +1,640 @@
+/*
+ * Copyright (c) 2021-2024 ATL Electronics
+ * Copyright (c) 2024 MASSDRIVER EI (massdriver.space)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT bflb_bl_uart
+
+/**
+ * @brief UART driver for Bouffalo Lab MCU family
+ */
+
+#include <zephyr/drivers/uart.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/arch/common/sys_io.h>
+#include <zephyr/drivers/timer/system_timer.h>
+#include <zephyr/pm/policy.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/irq.h>
+
+#include <glb_reg.h>
+#include <hbn_reg.h>
+#include <pds_reg.h>
+#include <bouffalolab/common/uart_reg.h>
+#include <bouffalolab/common/bflb_uart.h>
+
+/* clang-format off */
+
+struct bflb_config {
+	const struct pinctrl_dev_config *pincfg;
+	uint32_t baudrate;
+	uint8_t direction;
+	uint8_t data_bits;
+	uint8_t stop_bits;
+	uint8_t parity;
+	uint8_t bit_order;
+	uint8_t flow_ctrl;
+	uint8_t tx_fifo_threshold;
+	uint8_t rx_fifo_threshold;
+	uint32_t base_reg;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_config_func_t irq_config_func;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+struct bflb_data {
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t user_cb;
+	void *user_data;
+#endif /* CONFIG_UART_INTERRUPT_DRIVEN */
+};
+
+static void uart_bflb_enabled(const struct device *dev, uint32_t enable)
+{
+	const struct bflb_config *cfg = dev->config;
+	uint32_t rxt = 0;
+	uint32_t txt = 0;
+
+	if (enable) {
+		enable = 1;
+	}
+
+	txt = sys_read32(cfg->base_reg + UART_UTX_CONFIG_OFFSET);
+	txt = (txt & ~UART_CR_UTX_EN) | enable;
+	rxt = sys_read32(cfg->base_reg + UART_URX_CONFIG_OFFSET);
+	rxt = (rxt & ~UART_CR_URX_EN) | enable;
+	sys_write32(rxt, cfg->base_reg + UART_URX_CONFIG_OFFSET);
+	sys_write32(txt, cfg->base_reg + UART_UTX_CONFIG_OFFSET);
+}
+
+static uint32_t uart_bflb_get_crystal_frequency(void)
+{
+	uint32_t tmpVal;
```

Comment:
Let's not use camelcase (here and in the rest of the changes)

Reply:
It's in the inherited coding style described here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style

Reply:
Also, there should not be "a lot" of CamelCase code in Zephyr. Please point to an example of this.

Reply:
> Also, there should not be "a lot" of CamelCase code in Zephyr. Please point to an example of this.
random pick: i2c_imx.c, note im not talking about the SDK calls in this file, but for example 'i2cInitConfig' or 'rxBuffer'.

> It's in the inherited coding style described here: https://docs.zephyrproject.org/latest/contribute/guidelines.html#coding-style

Alright, but It does not forbid it ('Frowned upon' is the only opinion expressed about it), rather, it advises against bad variable naming, not variable naming style, which I'd argue the example given is a horrible example of (temporary counter as tmp? Should be i, or cnt). Renaming it to tmp_val would go against those guidelines, just tmp could work.

Reply:
> random pick: i2c_imx.c, note im not talking about the SDK calls in this file, but for example 'i2cInitConfig' or 'rxBuffer'.

Okay, those goes against our established coding style.

> Alright, but It does not forbid it ('Frowned upon' is the only opinion expressed about it), rather, it advises against bad variable naming, not variable naming style, which I'd argue the example given is a horrible example of (temporary counter as tmp? Should be i, or cnt). Renaming it to tmp_val would go against those guidelines, just tmp could work.

Nevertheless, please avoid using CamelCase in Zephyr.

Reply:
OK. It could maybe use a mention in the guidelines (something simple like 'Camelcase and mixed-case variable naming should be avoided entirely' maybe? using both ways to call it so ctrl+f or google brings it up) To express it's not just frowned upon and clarify the situation, as looking at the existing code didn't.

Reply:
Fixed